contracts/cw1155-base/src/contract.rs

```diff
@@ -0,0 +1,614 @@
+};
+use cw2::set_contract_version;
+
+use crate::error::ContractError;
+use crate::msg::InitMsg;
+use crate::state::{APPROVES, MINTER, TOKENS};
+
+// version info for migration info
+const CONTRACT_NAME: &str = "crates.io:cw1155-base";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    msg: InitMsg,
+) -> StdResult<Response> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+    let minter = deps.api.canonical_address(&msg.minter)?;
+    MINTER.save(deps.storage, &minter)?;
+    Ok(Response::default())
+}
+
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: Cw1155HandleMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        Cw1155HandleMsg::TransferFrom {
+            from,
+            to,
+            token_id,
+            value,
+        } => execute_transfer_from(deps, info, from, to, token_id, value, None),
+        Cw1155HandleMsg::SendFrom {
+            from,
+            contract,
+            token_id,
+            value,
+            msg,
+        } => execute_transfer_from(deps, info, from, Some(contract), token_id, value, Some(msg)),
+        Cw1155HandleMsg::BatchTransferFrom { from, to, batch } => {
+            execute_batch_transfer_from(deps, env, info, from, to, batch, None)
+        }
+        Cw1155HandleMsg::BatchSendFrom {
+            from,
+            contract,
+            batch,
+            msg,
+        } => execute_batch_transfer_from(deps, env, info, from, contract, batch, Some(msg)),
+        Cw1155HandleMsg::SetApprovalForAll { operator, approved } => {
+            execute_set_approval(deps, env, info, operator, approved)
+        }
+    }
+}
+
+fn execute_transfer_from_inner(
+    deps: &mut DepsMut,
+    info: &MessageInfo,
+    from: Option<HumanAddr>,
+    to: Option<HumanAddr>,
+    token_id: TokenId,
+    amount: Uint128,
+    msg: Option<Option<Binary>>,
+) -> Result<Response, ContractError> {
+    if let Some(from) = &from {
+        let canonical_from = deps.api.canonical_address(from)?;
+        TOKENS.update(
+            deps.storage,
+            (&token_id, canonical_from.as_slice()),
+            |balance: Option<Uint128>| balance.unwrap_or_default() - amount,
+        )?;
+    }
+
+    if let Some(to) = &to {
+        let canonical_to = deps.api.canonical_address(to)?;
+        TOKENS.update(
+            deps.storage,
+            (&token_id, canonical_to.as_slice()),
+            |balance: Option<Uint128>| -> StdResult<_> { Ok(balance.unwrap_or_default() + amount) },
+        )?;
+    }
+
+    let messages = match msg {
+        Some(msg) => vec![Cw1155ReceiveMsg {
+            operator: info.sender.clone(),
+            from: from.clone(),
+            amount,
+            token_id: token_id.clone(),
+            msg,
+        }
+        .into_cosmos_msg(to.clone().unwrap())?], // `to` in Send messages must not be None
```

Comment:
not sure if [MIRAI contracts](https://github.com/facebookexperimental/MIRAI#using-mirai-together-with-the-rust-design-by-contracts-crate) can be used to express this?